src/main/java/org/terasology/launcher/game/Installation.java

```diff
@@ -70,35 +87,50 @@ public String toString() {
                 .toString();
     }
 
+    static Path findJar(Path searchPath, BiPredicate<Path, BasicFileAttributes> predicate, String displayName) throws IOException {
+        Path resultPath;
+
+        try (var jarMatches = Files.find(searchPath, 3, predicate)) {
+            var matches = jarMatches.collect(Collectors.toUnmodifiableSet());
+            if (matches.isEmpty()) {
+                throw new FileNotFoundException("Could not find " + displayName + " jar in " + searchPath);
+            } else if (matches.size() > 1) {
+                throw new FileNotFoundException(
+                        String.format("Ambiguous results while looking for " + displayName + " jar in %s: %s",
+                                searchPath, matches));
+            }
+            resultPath = matches.iterator().next();
+        }
+        return resultPath;
+    }
+
     /**
      * Find the version of the Terasology engine installed here.
      * <p>
      * This method assumes that there is exactly one file named {@code engine*.jar} in the installation.
      *
      * @param versionDirectory the location containing the installation
-     * @throws IOException if an I/O error occurs
+     * @throws IOException           if an I/O error occurs
      * @throws FileNotFoundException if the engine or the version info could not be found
      */
     static Semver getEngineVersion(Path versionDirectory) throws IOException {
-        Path engineJar;
-
-        try (var jarMatches = Files.find(versionDirectory, 3, Installation::matchEngineJar)) {
-            var matches = jarMatches.collect(Collectors.toUnmodifiableSet());
-            if (matches.isEmpty()) {
-                throw new FileNotFoundException("Could not find engine jar in " + versionDirectory);
-            } else if (matches.size() > 1) {
-                throw new FileNotFoundException(
-                        String.format("Ambiguous results while looking for engine jar in %s: %s",
-                                versionDirectory, matches));
-            }
-            engineJar = matches.iterator().next();
-        }
-
+        Path engineJar = findJar(versionDirectory, Installation::matchEngineJar, "engine");
         Properties versionInfo = getVersionPropertiesFromJar(engineJar);
-
         return new Semver(versionInfo.getProperty("engineVersion"), Semver.SemverType.IVY);
     }
 
+    static BiPredicate<Path, BasicFileAttributes> matchJar(Predicate<String> predicate) {
+        return (path, basicFileAttributes) -> {
```

Comment:
This one could use some documentation. Takes one kind of predicate, returns a different one.

Specifically, in this current implementation, the argument of this `predicate` is the **string form** of the **last component** of the path.

and it operates on a String because you refactored this from the code that was using `file.endsWith(".jar") && file.startsWith("engine")`, which needed strings. (Arguably the test for `Terasology.jar` would do just as well being a Path, because it doesn't involve substring comparison. Is Path comparison case-insensitive on case-insensitive filesystems?)

that was a tangent — I think it's a toss-up as to whether that should take String or Path — my main point was that we need to document what we're passing to those predicates.

Reply:
:books: :newspaper:  :nerd_face:

Reply:
Haven't forgotten about this... will get to it eventually!

Reply:
I think that one will still take me a bit to wrap my head around the next time I come across it, but those docs should help. :+1: